src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs

```diff
@@ -0,0 +1,360 @@
+            // before attaching proxy credentials and marking the HTTP request to be re-submitted.
+            // So we need to reset the LastStatusCode remembered. Otherwise, it will see additional 407
+            // responses as an indication that proxy auth failed and won't retry the HTTP request.
+            if (state.LastStatusCode == HttpStatusCode.ProxyAuthenticationRequired)
+            {
+                state.LastStatusCode = 0;
+            }
+
+            // For security reasons, we drop the server credential if it is a 
+            // NetworkCredential.  But we allow credentials in a CredentialCache
+            // since they are specifically tied to URI's.
+            if (!(state.ServerCredentials is CredentialCache))
+            {
+                state.ServerCredentials = null;
+            }
+        }
+        
+        private static void OnRequestSendingRequest(WinHttpRequestState state)
+        {
+            Debug.Assert(state != null, "OnRequestSendingRequest: state is null");
+            
+            if (state.RequestMessage.RequestUri.Scheme != WinHttpRequestState.UriSchemeHttps)
+            {
+                // Not SSL/TLS.
+                return;
+            }
+
+            // Grab the channel binding token (CBT) information from the request handle and put it into
+            // the TransportContext object.
+            state.TransportContext.SetChannelBinding(state.RequestHandle);
+
+            if (state.ServerCertificateValidationCallback != null)
+            {
+                IntPtr certHandle = IntPtr.Zero;
+                uint certHandleSize = (uint)IntPtr.Size;
+
+                if (!Interop.WinHttp.WinHttpQueryOption(
+                    state.RequestHandle,
+                    Interop.WinHttp.WINHTTP_OPTION_SERVER_CERT_CONTEXT,
+                    ref certHandle,
+                    ref certHandleSize))
+                {
+                    int lastError = Marshal.GetLastWin32Error();
+                    throw WinHttpException.CreateExceptionUsingError(lastError);
+                }
+                
+                // Create a managed wrapper around the certificate handle. Since this results in duplicating
+                // the handle, we will close the original handle after creating the wrapper.
+                var serverCertificate = new X509Certificate2(certHandle);
+                Interop.Crypt32.CertFreeCertificateContext(certHandle);
+
+                X509Chain chain = null;
+                SslPolicyErrors sslPolicyErrors;
+
+                try
+                {
+                    WinHttpCertificateHelper.BuildChain(
+                        serverCertificate,
+                        state.RequestMessage.RequestUri.Host,
+                        state.CheckCertificateRevocationList,
+                        out chain,
+                        out sslPolicyErrors);
+
+                    bool result = state.ServerCertificateValidationCallback(
```

Comment:
I think we should always enable the flag, It's critical for best performance to avoid thread switching. And we shouldn't penalize devs that need to have a `ServerCertificateValidationCallback` method. 

We can document that the user code needs to follow similar rules about not blocking and returning quickly.